src/sagas/pushNotification.js

```diff
@@ -0,0 +1,432 @@
+ * @example
+ * {
+ *   bodyLocArgs: '[\"10 T2\",\"5 T1\",\"2\"]',
+ *   bodyLocKey: 'new_transaction_received_description_with_tokens',
+ *   titleLocKey: 'new_transaction_received_title',
+ *   txId: 'txId1',
+ * }
+ * @inner
+ */
+const messageHandler = async (message) => {
+  const { data } = message;
+  if (!localization.hasKey(data.titleLocKey)) {
+    console.log('unknown message titleLocKey', data.titleLocKey);
+    return;
+  }
+
+  if (!localization.hasKey(data.bodyLocKey)) {
+    console.log('unknown message bodyLocKey', data.bodyLocKey);
+    return;
+  }
+
+  const bodyArgs = JSON.parse(data.bodyLocArgs);
+  const title = localization.getMessage(data.titleLocKey);
+  const body = localization.getMessage(data.bodyLocKey, bodyArgs);
+
+  notifee.displayNotification({
+    title,
+    body,
+    android: {
+      channelId: TRANSACTION_CHANNEL_ID,
+    },
+  });
+};
+
+const showPinScreenForResult = async (dispatch) => new Promise((resolve) => {
+  const params = {
+    cb: (_pin) => {
+      dispatch(setIsShowingPinScreen(false));
+      resolve(_pin);
+    },
+    canCancel: false,
+    screenText: t`Enter your 6-digit pin to authorize operation`,
+    biometryText: t`Authorize operation`,
+  };
+
+  NavigationService.navigate('PinScreen', params);
+
+  // We should set the global isShowingPinScreen
+  dispatch(setIsShowingPinScreen(true));
+});
+
+/**
+ * This function is called when the wallet is initialized with success.
+ */
+export function* onAppInitialization() {
+  yield take(types.START_WALLET_SUCCESS);
+
+  const { enabled, showAmountEnabled } = STORE.getItem(pushNotificationKey.settings);
+  const hasBeenEnabled = STORE.getItem(pushNotificationKey.hasBeenEnabled);
+  const enabledAt = STORE.getItem(pushNotificationKey.enabledAt);
+
+  const persistedDeviceId = STORE.getItem(pushNotificationKey.deviceId);
+  const deviceId = yield call(getDeviceId);
+  // If the deviceId is different from the persisted one, we should update it
+  if (persistedDeviceId && persistedDeviceId !== deviceId.toString()) {
+    STORE.setItem(pushNotificationKey.deviceId, deviceId);
+    yield put(pushUpdateDeviceId({ deviceId }));
+  }
+
+  // Initialize the pushNotification state on the redux store
+  yield put(pushInit({
+    deviceId,
+    settings: {
+      enabled,
+      showAmountEnabled,
+    },
+    hasBeenEnabled,
+    enabledAt,
+  }));
+
+  // Create the transaction channel if it doesn't exist
+  const hasTransactionChannel = yield call(notifee.isChannelCreated, TRANSACTION_CHANNEL_ID);
+  if (!hasTransactionChannel) {
+    yield call(notifee.createChannel, {
+      id: TRANSACTION_CHANNEL_ID,
+      name: TRANSACTION_CHANNEL_NAME,
+    });
+  }
+
+  // Make sure deviceId is registered on the FCM
+  messaging().registerDeviceForRemoteMessages();
+  // Add listeners for push notifications on foreground and background
+  messaging().onMessage(onForegroundMessage);
+  messaging().setBackgroundMessageHandler(onBackgroundMessage);
```

Comment:
Can any of those methods fail and break the wallet?

Reply:
Yes, it can fail. I will wrap then in a try/catch.